### DISCUSSION TITLE
How to improve search?

### DISCUSSION THREAD

**Discussion by Gogex1** [MilesCranmer/PySR#242] [Q&A]
My model seems to be stuck, as for 1400 iterations it didn't change hall of fame at all. What parameters can i change to improve it?

**Comment by MilesCranmer** [comment]
It really depends on the problem, but have a look through the options:
https://astroautomata.com/PySR/api/#pysrregressor-parameters

In particular the search space: https://astroautomata.com/PySR/api/#creating-the-search-space

and the constraints: https://astroautomata.com/PySR/api/#working-with-complexities 

are going to be key to tuning. 

The search size parameters are for making the search larger, which is also necessary for more exhaustive searches: https://astroautomata.com/PySR/api/#setting-the-search-size

**Reply by Gogex1** [reply]
I've tried setting these, but it doesn't really help. Is it possible that it will find better solution if i leave it searching for a few days ?

**Reply by MilesCranmer** [reply]
Which ones did you try?

**Reply by DenisSvirin** [reply]
constraints/ no constraints
and population (20, 100, 1000)

**Reply by MilesCranmer** [reply]
Wait are you the same person? If so, which constraints?

**Reply by MilesCranmer** [reply]
It would be helpful if you put up a GitHub gist of your full search script

**Reply by DenisSvirin** [reply]
Oh i forgot to change account on my computer. Ye its my account.
i added pow(1,-1) for equation to be more physically accurate

**Reply by DenisSvirin** [reply]
https://gist.github.com/DenisSvirin/8bf35464e3e4a681a0d7bb4e2902149f

**Reply by MilesCranmer** [reply]
Comments:
- Don't use `neg` since you already have `-` as a binary operator. The fewer operators the better.
- Don't use `exp` since you are already using `^` (same reason as above).
- Use `/` instead of `inv` since there are some simplification utilities available for `/`.
- I think the constraint on `pow_abs` is backwards; you actually want it to be `(-1, 1)` (any base, only constant exponents).
- Try setting up a constraint on `*` as well.
- Why tweak `weight_randomize`?

**Reply by DenisSvirin** [reply]
i thought, that if i increase weight_randomize, i will, potentially, find more suitable equation

**Reply by DenisSvirin** [reply]
Also if im going to change evaluation function ( and derivative). Is there anything else i need to change for proper algorithm performance?

**Reply by MilesCranmer** [reply]
There isn't a general answer for this; it is problem dependent and hard to know what values are optimal in advance. You could try to do a tuning with, e.g., [GridSearchCV](https://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) or wandb. Personally I use wandb for tuning parameters. Here is the sweep file I have used in the past: https://github.com/MilesCranmer/pysr_wandb/blob/master/sweep.yml where the performance is calculated with this function: https://github.com/MilesCranmer/pysr_wandb/blob/348964dcdc153ded6eb624481e8eeaf9d1e3f62f/pysr_wandb/pysr_wandb.py#L140 which takes the median over a variety of seeds and problems.

**Reply by MilesCranmer** [reply]
A good rule of thumb is to minimize the possible search space (https://astroautomata.com/PySR/api/#creating-the-search-space), use as many constraints as you can (https://astroautomata.com/PySR/api/#working-with-complexities), and maximize the search size (https://astroautomata.com/PySR/api/#setting-the-search-size). The other parameters you need to tune.
